### PR TITLE
hevm: rpc cache

### PR DESCRIPTION
**Pull request by rainbreak**
rpc cache for dapp

```
  dapp --make-cache /tmp/dapp.cache 
  dapp test --rpc --cache /tmp/dapp.cache
```

### CODE REVIEW

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 1]
```diff
@@ -355,15 +363,32 @@ findJsonFile Nothing = do
...
+          Nothing ->
+            pure ()
+          Just path ->
+            -- merge all of the post-vm caches and save into the state
+            let
+              cache' = mconcat [view EVM.cache vm | vm <- vms]
+            in
+              Git.saveFacts (Git.RepoAt path) (Facts.cacheFacts cache')
+
+        when (any (== False) passing) exitFailure
```
Comment (MrChico) [inline src/hevm/hevm-cli/hevm-cli.hs:391]:
```suggestion
        unless (all passing) exitFailure
```

File: src/hevm/src/EVM/UnitTest.hs [hunk 2]
```diff
@@ -450,20 +459,22 @@ runUnitTestContract
...
-          let bailing = [x | Left x <- details]
+          let running = [x | (Right x, _) <- details]
+          let bailing = [x | (Left  x, _) <- details]
 
           tick "\n"
           tick (Text.unlines (filter (not . Text.null) running))
           tick (Text.unlines (filter (not . Text.null) bailing))
 
-          pure (null bailing)
+          let isRight (Right _) = True
```
Comment (MrChico) [inline src/hevm/src/EVM/UnitTest.hs:475]: better to import from Data.Either

**Comment by rainbreak** [comment]
@MrChico I didn't add `saveFacts` at the end of `symbolic` as I couldn't really tell what was going on. I'll leave that one to you. All finished otherwise I think.

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 3]
```diff
@@ -226,6 +232,24 @@ instance Options.ParseRecord (Command Options.Wrapped) where
 optsMode :: Command Options.Unwrapped -> Mode
 optsMode x = if debug x then Debug else Run
 
+applyCache :: (Maybe String, Maybe String) -> IO (EVM.VM -> EVM.VM)
+applyCache (state, cache) =
+  let applyState = do flip Facts.apply
+      applyCache = do flip Facts.applyCache
```
Comment (MrChico) [inline src/hevm/hevm-cli/hevm-cli.hs:238]:
```suggestion
  let applyState = flip Facts.apply
      applyCache = flip Facts.applyCache
```

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 4]
```diff
@@ -48,6 +48,7 @@ import qualified EVM.UnitTest
 import Control.Concurrent.Async   (async, waitCatch)
 import Control.Lens hiding (pre)
 import Control.Monad              (void, when, forM_, unless)
+import Control.Monad.IO.Class     (liftIO)
```
Comment (MrChico) [inline src/hevm/hevm-cli/hevm-cli.hs:51]:
```suggestion
```

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 5]
```diff
@@ -743,14 +787,15 @@ symvmFromCommand cmd = do
     Just SymbolicS -> EVM.Symbolic <$> freshArray_ Nothing
     Nothing -> EVM.Symbolic <$> freshArray_ (if create cmd then (Just 0) else Nothing)
 
+  withCache <- liftIO $ applyCache (state cmd, cache cmd)
```
Comment (MrChico) [inline src/hevm/hevm-cli/hevm-cli.hs:790]:
```suggestion
  withCache <- io $ applyCache (state cmd, cache cmd)
```

File: src/hevm/hevm-cli/hevm-cli.hs (hunk 4 above)
Comment (MrChico) [reply src/hevm/hevm-cli/hevm-cli.hs:51]: there's a specialized `io :: (MonadQuery m) => IO a -> m a` included from SBV already

**Review by MrChico — APPROVED** [review]

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 6]
```diff
@@ -355,15 +372,32 @@ findJsonFile Nothing = do
...
-        when (any (== False) results) exitFailure
+
+        results <- concatMapM (runUnitTestContract opts contractMap sourceCache) unitTests
+        let (passing, vms) = unzip results
+
+        case cache of
+          Nothing ->
+            pure ()
+          Just path ->
+            -- merge all of the post-vm caches and save into the state
```
Comment (MrChico) [inline src/hevm/hevm-cli/hevm-cli.hs:394]: If the cache was shared between unitTests we'd have better execution for the initial run and runs withoug `--cache`

Comment (rainbreak) [reply src/hevm/hevm-cli/hevm-cli.hs:394]: Yes, agree, was just a bit more work to implement.

**Pull request merged**